### PR TITLE
General: Connections: Add MobileMagicLink component

### PR DESCRIPTION
**Pull request by ebinnion**
#### Changes proposed in this Pull Request:

* Adds ability to send magic link from Jetpack to log user in to mobile app.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* This adds new functionality.
* See p9OQAC-eq-p2. 

#### Testing instructions:

Testing the error case:

- Ensure that Jetpack site is connected to a test account that **is** an a11n account
- Go to Jetpack > Dashboard
- Click **Connect to mobile WordPress app** link
- Ensure modal pops up
- Click **Send Link** button
- Ensure that an error message occurs (this is due to you being connected to an a11n account)
- Disconnect site

Testing the success case: 
- Reconnect site to a WordPress.com test user that is not your a11n account
- Click **Connect to mobile WordPress app** link
- Ensure modal pops up
- Click **Send Link** button
- Ensure that you receive email with magic link

#### Design Changes:

Before this PR, when going to the Jetpack Dashboard you would see a plain "Account Connection" card under the heading "Connections":

Screen Shot 2019-06-06 at 10 09 28

This PR adds the blue link in your "Account Connection" card that says: **Connect to mobile WordPress app**:

Screen Shot 2019-06-06 at 10 08 43

This link opens up a this modal:

Screen Shot 2019-06-06 at 10 09 00

Clicking "Cancel" or outside the modal area will dismiss it. Clicking on "Send link" will issue a request to the Jetpack backend which in turn calls the WPCOM backend. This should result in an email being sent with a link that allows you to log in to the mobile app.

As the request is being made, the following notice is displayed in the upper right corner:

Screen Shot 2019-06-06 at 10 12 11

Upon success, this notice is displayed:

Screen Shot 2019-06-06 at 10 12 01

Upon error, this notice is displayed:

Screen Shot 2019-06-06 at 10 40 34

#### Proposed changelog entry for your changes:

* Admin Dashboard: Add a way for users to get a magic link to log into the mobile app.

### CODE REVIEW

**Comment by poisa** [comment]
@joanrho You are welcome to look at this!

File: _inc/client/components/mobile-magic-link/index.jsx [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+					onKeyDown={ onKeyDownCallback( this.handleOpenModal ) }
+					role="button"
+					tabIndex="0"
+				>
+					{ __( 'Connect to mobile WordPress app' ) }
+				</a>
+				{ showModal && (
+					<Modal className="mobile-magic-link__modal" onRequestClose={ this.toggleModalVisibility }>
+						<Card className="mobile-magic-link__modal__body">
+							<h2>{ __( 'Email me a link to log in the app' ) }</h2>
```
Comment (aerych) [inline _inc/client/components/mobile-magic-link/index.jsx]: Might be missing a "to" in `Email me a link to log in the app`. 
`Email me a link to log in to the app`

File: _inc/client/components/mobile-magic-link/index.jsx [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+							<h4>
+								{ __(
+									"Easily log in to the WordPress.com app by clicking the link we'll send to the email address on your account."
```
Comment (aerych) [inline _inc/client/components/mobile-magic-link/index.jsx]: `Easily log in to the WordPress.com app by clicking [...]`
Probably we don't want to say "WordPress.com".  How about just "WordPress".

Comment (aerych) [reply _inc/client/components/mobile-magic-link/index.jsx]: Maybe:
"Easily log in to the WordPress mobile app with a special link sent to your email address.  Just click the link from your mobile device to log in!"  

Dunno. Words are hard.

**Review by aerych — COMMENTED** [review]
Left a couple of thoughts about the copy.

File: _inc/client/components/mobile-magic-link/index.jsx [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+		return (
+			<div className="mobile-magic-link">
+				<a
+					className="mobile-magic-link__button"
+					onClick={ this.handleOpenModal }
+					onKeyDown={ onKeyDownCallback( this.handleOpenModal ) }
+					role="button"
+					tabIndex="0"
+				>
+					{ __( 'Connect to mobile WordPress app' ) }
```
Comment (aerych) [inline _inc/client/components/mobile-magic-link/index.jsx]: Maybe "Log in to the WordPress mobile app" if there is enough horizontal room in the card? 
Maybe swap "WordPress" and "mobile" regardless? Seems like we tend to say "WordPress mobile app" more naturally than the other way. (Or maybe that's just me. 🤷‍♂ )

**Review by jeherve — COMMENTED** [review]
This worked well for me, it's handy. 

I was a bit surprised by the location of the link, though. It's really hidden for such a useful feature.
It also gets to be a bit noisy for secondary users. The 2 links are right below each other:

image

I also wonder if the modal's design should look like the other modals used in the past inside the Jetpack dashboard? Those designs also included a "close" icon, for example, and different font sizes and colors.

File: _inc/client/components/mobile-magic-link/index.jsx (hunk 2 above)
Comment (ebinnion) [reply _inc/client/components/mobile-magic-link/index.jsx]: > Dunno. Words are hard.

Heard that. 😂

Comment (poisa) [reply _inc/client/components/mobile-magic-link/index.jsx]: > Words are hard.

We (non native speakers) agree!

_97587522_9ea23dbd-7ff4-4228-9f5c-94824ed857fc

**Comment by joanrho** [comment]
Due to some issue with my local setup, I was only able to test with a Jurassic Ninja test site + beta plugin spinning up this branch: the text link, modal, and error message all worked as expected. 

However, everything here in the instructions and below, "Check out D29138-code on WPCOM sandbox", I was unable to test. I then pinged @MichaelArestad to test, and he was able to test the beginning part on his sandbox site succesfully, then wasn't able to get the WPCOM sandbox part to work.

Would it be possible to make the testing instructions a bit easier (for non-devs) or perhaps record a video of the process to test that latter part?

**Comment by poisa** [comment]
> then wasn't able to get the WPCOM sandbox part to work.

Yes, that bit actually caught me off guard too as I was working on this. I have added more details in the testing instructions with a link to the FG where it goes into much more detail than I have. 

>Would it be possible to make the testing instructions a bit easier (for non-devs) or perhaps record a >video of the process to test that latter part?

@joanrho, see if the additional instructions I added help in any way. If not I'll gladly make a video with the process 😄

**Comment by ebinnion** [comment]
I've deployed D29138-code on WPCOM and have updated the testing instructions here accordingly.

**Comment by MichaelArestad** [comment]
Much easier to test, now.

The only thing that really stuck out to me was where the "Connect to mobile WordPress app" is located. Because I don't have much context in how this might be used, I can't make any specific suggestions yet.

Why would a user want to click this link? Where are they coming from? If they're using the mobile app, would they need to be next to a computer?

**Comment by ebinnion** [comment]
> Why would a user want to click this link? 

To be more easily logged in to the WordPress mobile app by simply clicking a link, as opposed to having to enter username, password, etc.

> Where are they coming from? 

I'm not sure I have a great answer for this. cc @aerych who might've put more thought into it than I.

> If they're using the mobile app, would they need to be next to a computer?

I think the idea here is that we're trying to get users that aren't yet using the mobile app to more easily get started in the mobile app. But, I'll defer to @aerych on that as well.

**Comment by ebinnion** [comment]
My personal preference was for the link to go in the mobile apps prompt that shows at the bottom the Jetpack dashboard:

Screen Shot 2019-06-07 at 4 03 04 PM

But, I got some early feedback that the link may go better in the account section for a couple of reasons:

1) The mobile banner is dismissible. So, once a user dismisses it, there's not a way for the user to change their mind and use the magic link.
2) The link in the account section seemed like it'd be easier to iterate on

**Comment by joanrho** [comment]
@ebinnion thanks for the updated testing instructions—was able to test all the way to the end of the instructions this time. Everything worked as expected. Minor copy nit though:

>Email me a link to log in the app

should probably be "Email me a link to log in**to** the app"

@MichaelArestad @ebinnion Great points above about where the best place for this link should be. I think the working solution is the best to start with in addition to Eric's proposed link update to the mobile apps banner at the bottom of the Jetpack Dashboard.

**Comment by aerych** [comment]
> > Why would a user want to click this link?
> 
> To be more easily logged in to the WordPress mobile app by simply clicking a link, as opposed to having to enter username, password, etc.

Yep yep. For some reason there are folks who struggle with credentials. Tapping a link helps to stream line the login process. 

> > Where are they coming from?
> 
> I'm not sure I have a great answer for this. cc @aerych who might've put more thought into it than I.

To some degree, organic discovery. Others would be directed either through interactions with happiness engineers or tbd future marketing / communication channels.

> > If they're using the mobile app, would they need to be next to a computer?
> 
> I think the idea here is that we're trying to get users that aren't yet using the mobile app to more easily get started in the mobile app. But, I'll defer to @aerych on that as well.

Exactly this. 

@elibud anything you'd add?

**Comment by elibud** [comment]
I don't think this is a bad place to add it, but[ I agree with @aerych ](https://github.com/Automattic/jetpack/pull/12561#discussion_r291402958)that "Connect to mobile WordPress app" doesn't really convey what's going on here.

Also, shouldn't the link on the banner mentioned by @ebinnion [here](https://github.com/Automattic/jetpack/pull/12561#issuecomment-500038412) trigger the same flow?

**Comment by aerych** [comment]
> should probably be "Email me a link to log in**to** the app"

Nitpicky but for consistency with other [nux copy on WordPress.com](https://wordpress.com/log-in?redirect_to=https%3A%2F%2Fwordpress.com%2F) and in the apps I think we'd want "in to" vs "into". 
Screen Shot 2019-06-10 at 9 43 44 AM

**Comment by ebinnion** [comment]
> Also, shouldn't the link on the banner mentioned by @ebinnion here trigger the same flow?

Not currently. That was a suggestion for an alternative, which we may do in the future.

> "Connect to mobile WordPress app" doesn't really convey what's going on here.

Do y'all have a preference for the phrase to use here? cc @aerych @elibud 

> should probably be "Email me a link to log into the app"

Will do 👍

**Comment by ebinnion** [comment]
I just pushed a commit that handles changing the modal header to be in line with @aerych's suggestion.

Screen Shot 2019-06-10 at 10 18 56 AM

**Comment by ebinnion** [comment]
[[4] Magic Links in Jetpack UI](https://trello.com/c/8MKNM6LY/251-4-magic-links-in-jetpack-ui)

**Comment by aerych** [comment]
> > "Connect to mobile WordPress app" doesn't really convey what's going on here.
> 
> Do y'all have a preference for the phrase to use here? cc @aerych @elibud

I'd left a comment about it here https://github.com/Automattic/jetpack/pull/12561#discussion_r291402958

**Comment by ebinnion** [comment]
@Automattic/editorial – Could y'all please provide some assistance with copy?

It seems like the main question is around the link that pops the modal up. Right now, it says, "Connect to mobile WordPress app"

One suggestion is to switch that to "Log in to the WordPress mobile app".

For context, here is where that link is displayed:

https://github.com/Automattic/jetpack/pull/12561#pullrequestreview-247044437

**Comment by michelleweber** [comment]
"Log in to the WordPres mobile app" seems the most straightforward and clear, I'd go with that.

**Comment by ebinnion** [comment]
Link has been updated. Thanks for the copy suggestion!

Screen Shot 2019-06-13 at 2 34 00 PM

@jeherve @aerych  I think we're ready for a code review now 🤞

File: _inc/client/components/mobile-magic-link/index.jsx [hunk 4]
```diff
@@ -55,7 +55,7 @@ export class MobileMagicLink extends React.Component {
 					role="button"
 					tabIndex="0"
 				>
-					{ __( 'Connect to mobile WordPress app' ) }
+					{ __( 'Log in to the WordPres mobile app' ) }
```
Comment (aerych) [inline _inc/client/components/mobile-magic-link/index.jsx]: I think the last "s" in WordPress managed to escape when no one was looking.

**Review by jeherve — DISMISSED** [review]
This looks good to me, and tests well. I left 2 comments, but I don't think any of them are blocking. I think they could be addressed in a future PR. I'll let you judge.

File: _inc/client/components/mobile-magic-link/index.jsx [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+	toggleModalVisibility = () => {
+		this.setState( {
+			showModal: ! this.state.showModal,
+		} );
+	};
+
+	clickSendLoginEmail = e => {
+		e.preventDefault();
+		this.toggleModalVisibility();
+		this.props.sendMobileLoginEmail();
```
[the remark this answers is not in the document]
Comment (ebinnion) [reply _inc/client/components/mobile-magic-link/index.jsx]: I've pushed a change which address that 👍

File: _inc/client/components/mobile-magic-link/index.jsx [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+					className="mobile-magic-link__button"
+					onClick={ this.handleOpenModal }
+					onKeyDown={ onKeyDownCallback( this.handleOpenModal ) }
+					role="button"
+					tabIndex="0"
+				>
+					{ __( 'Log in to the WordPress mobile app' ) }
+				</a>
+				{ showModal && (
+					<Modal className="mobile-magic-link__modal" onRequestClose={ this.toggleModalVisibility }>
```
[the remark this answers is not in the document]
Comment (ebinnion) [reply _inc/client/components/mobile-magic-link/index.jsx]: Done 😄

**Comment by ebinnion** [comment]
@jeherve Thank you for the feedback. I've pushed a commit that addresses your feedback.

**Review by jeherve — APPROVED** [review]
This looks good to me. Merge when ready!

**Pull request merged**